lib/datadog/tracing/contrib/pg/instrumentation.rb

```diff
@@ -0,0 +1,75 @@
+# typed: false
+
+require 'datadog/tracing'
+require 'datadog/tracing/metadata/ext'
+require 'datadog/tracing/contrib/analytics'
+require 'datadog/tracing/contrib/pg/ext'
+
+module Datadog
+  module Tracing
+    module Contrib
+      module Pg
+        # PG::Connection patch module
+        module Instrumentation
+          def self.included(base)
+            base.prepend(InstanceMethods)
+          end
+
+          # PG::Connection patch methods
+          module InstanceMethods
+            def exec(sql)
+              service = Datadog.configuration_for(self, :service_name) || datadog_configuration[:service_name]
+
+              Tracing.trace(Ext::SPAN_EXEC, service: service) do |span|
+                span.resource = sql
+                span.type = Tracing::Metadata::Ext::SQL::TYPE
+                span.service = Ext::DEFAULT_PEER_SERVICE_NAME
+
+                span.set_tag('db.instance', db)
+                span.set_tag('db.user', user)
+                span.set_tag('db.system', Ext::SPAN_SYSTEM)
+                span.set_tag('db.statement', sql)
+
+                span.set_tag('network.destination.name', host)
+                span.set_tag('network.destination.port', port)
+
+                span.set_tag(Tracing::Metadata::Ext::TAG_COMPONENT, Ext::TAG_COMPONENT)
+                span.set_tag(Tracing::Metadata::Ext::TAG_OPERATION, Ext::TAG_OPERATION_QUERY)
+
+                span.set_tag(Ext::TAG_DB_NAME, db)
+
+                # Tag as an external peer service
+                span.set_tag(Tracing::Metadata::Ext::TAG_PEER_SERVICE, service)
+                span.set_tag(Tracing::Metadata::Ext::TAG_PEER_HOSTNAME, host)
+                span.set_tag(Tracing::Metadata::Ext::NET::TAG_TARGET_HOST, host)
+                span.set_tag(Tracing::Metadata::Ext::NET::TAG_TARGET_PORT, port)
+
+                # Set analytics sample rate
+                Contrib::Analytics.set_sample_rate(span, analytics_sample_rate) if analytics_enabled?
+
+                result = super(sql)
+                span.set_tag('db.row_count', result.ntuples)
+                span.set_tag('sql.rows', result.ntuples)
```

Comment:
Do these operations read the whole result before returning their value?

If the their documentation isn't clear about it, you can try populating the test table with A LOT of records (anything that will make the query take many seconds) and see if these methods calls here still return super fast.

Reply:
I don't think `ntuples` requires reading the whole result.
I populated a table with ~2 million rows, and tried to get the number of rows returned from the `SELECT *` statement:
```
> conn.exec('SELECT * FROM test_table')
=> #<PG::Result:0x0000aaab21fdf490 status=PGRES_TUPLES_OK ntuples=2330066 nfields=2 cmd_tuples=2330066>
> _.ntuples
=> 2330066
```
the time it took to execute was around the same as the time it took for a table with very few rows.

Reply:
Awesome, keep it then. 👍